bentoml/_internal/utils/circus.py

```diff
@@ -61,94 +75,160 @@ def create_standalone_arbiter(watchers: list[Watcher], **kwargs: t.Any) -> Arbit
             )
 
 
-# TODO: use svc.build_args.include/exclude as default files to watch
-# TODO: watch changes in model store when "latest" model tag is used
-
-
-class BentoChangeReloader(CircusPlugin):
+class ReloaderPlugin(CircusPlugin):
     """
     A circus plugin that reloads the BentoService when the service code changes.
 
 
     Args:
         working_dir: The directory of the bento service.
-        reload_delay: The delay in seconds between checking for changes.
     """
 
     name = "bento_change_reloader"
     config: dict[str, t.Any]
 
+    @inject
     def __init__(self, *args: t.Any, **config: t.Any):
         assert "bento_identifier" in config, "`bento_identifier` is required"
         assert "working_dir" in config, "`working_dir` is required"
 
         super().__init__(*args, **config)  # type: ignore (unfinished types for circus)
 
         self.name = self.config.get("name")
+        self._cwd = Path.cwd()
+
         working_dir: str = self.config["working_dir"]
+        bentoml_home = self.config["bentoml_home"]
 
         # circus/plugins/__init__.py:282 -> converts all given configs to dict[str, str]
-        self.reload_delay: float = float(self.config.get("reload_delay", 1))
-        self.file_watcher = PyFileChangeWatcher([working_dir])
+        self.watcher_timeout: int = int(self.config.get("watcher_timeout", 0))
+
+        watch_dirs = [working_dir, fs.path.combine(bentoml_home, "models")]
+
+        if not is_pypi_installed_bentoml():
+            # git root from this __file__
+            logger.info(
+                "BentoML is installed via development mode, adding git root to `watch_dirs`"
+            )
+            git_root = str(Path(__file__).parent.parent.parent)
+            watch_dirs.append(git_root)
+
+        logger.info(f"Watching directories: {watch_dirs}")
+
+        self.exit_event = threading.Event()
+        self.filters = BentoFilter(self._cwd)
+
+        self.watch_dirs = watch_dirs
+
+        self.watchers = watchfiles.watch(
+            *self.watch_dirs,
+            watch_filter=None,
+            yield_on_timeout=True,  # to stop hanging on our tests, doesn't affect the behaviour
+            rust_timeout=self.watcher_timeout,  # no need to setup timeout
+            stop_event=self.exit_event,
+        )
+
+    def filtered(
```

Comment:
Is it possible to make use of the internal filter of watchfiles?

Reply:
we want to format the output correctly, hense we can't use the internal `watch_filter` here